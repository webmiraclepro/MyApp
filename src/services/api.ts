import axios from 'axios'
import get from 'lodash/get'
import map from 'lodash/map'
import { API_URL, API_KEY, API_TIMEOUT } from '../constants'
import { Photo } from '../models'

const getPhotos = async (): Promise<Photo[]> => {
  const response = await axios.get(`${API_URL}/${API_KEY}`, {
    params: {
      fmt: 'json',
    },
    timeout: API_TIMEOUT,
  })
  return map<any, Photo>(get(response, 'data.results[0]', []), (u, i) => ({
    ...u,
    id: String(i),
    photo: `https://picsum.photos/id/${i}/200/200`,
  }))
}

export default {
  getPhotos,
}
