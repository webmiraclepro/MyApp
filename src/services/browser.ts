import { Linking } from 'react-native'
import InAppBrowser from 'react-native-inappbrowser-reborn'

export default {
  openUrl: async (url: string) => {
    if (await InAppBrowser.isAvailable()) {
      InAppBrowser.open(url, {
        // iOS Properties
        dismissButtonStyle: 'cancel',
        preferredBarTintColor: '#453AA4',
        preferredControlTintColor: 'white',
        readerMode: false,
        animated: true,
        modalEnabled: true,
        // Android Properties
        showTitle: true,
        toolbarColor: '#6200EE',
        secondaryToolbarColor: 'black',
        enableUrlBarHiding: true,
        enableDefaultShare: true,
        forceCloseOnRedirection: false,
        // Specify full animation resource identifier(package:anim/name)
        // or only resource name(in case of animation bundled with app).
        animations: {
          startEnter: 'slide_in_right',
          startExit: 'slide_out_left',
          endEnter: 'slide_in_left',
          endExit: 'slide_out_right',
        },
      })
    } else {
      Linking.openURL(url)
    }
  },
}
