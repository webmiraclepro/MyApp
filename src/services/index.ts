export { default as ApiService } from './api'
export { default as NavigationService } from './navigation'
export { default as LogService } from './log'
export { default as BrowserService } from './browser'
export { default as PhotoService } from './photo'

export * from './database'
