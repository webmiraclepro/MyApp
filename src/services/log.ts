import { Client } from 'bugsnag-react-native'
import { BUGSNAG_KEY } from '../constants'
const bugsnag = new Client(BUGSNAG_KEY)

export default {
  logError: (err: Error): void => {
    bugsnag.notify(err)
  },
}
