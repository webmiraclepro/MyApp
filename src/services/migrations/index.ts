import SQLiteService from '../sqlite'
import LogService from '../log'
import migrationCreateTables from './createTables'
import { Migration } from './migration'

const TABLE_MIGRATION = 'migrations'

const migrations: Array<Migration> = [migrationCreateTables]

const applyMigration = async ({ id, queries }: Migration) => {
  try {
    await SQLiteService.executeSql(
      `INSERT INTO '${TABLE_MIGRATION}' ('id') VALUES (?);`,
      [id],
    )
    await SQLiteService.sqlBatch(queries)
  } catch (error) {
    LogService.logError(error)
  }
}

export const runMigrations = async () => {
  await SQLiteService.executeSql(
    `CREATE TABLE IF NOT EXISTS '${TABLE_MIGRATION}' (id TEXT PRIMARY KEY);`,
  )
  const migration: Array<Migration> = await SQLiteService.getItems(
    `SELECT * FROM ${TABLE_MIGRATION};`,
  )

  await Promise.all(
    migrations.map((newMigration) => {
      if (!migration.some(({ id }) => newMigration.id === id)) {
        return applyMigration(newMigration)
      }
      return Promise.resolve()
    }),
  )
}
