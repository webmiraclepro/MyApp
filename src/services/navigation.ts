import {
  NavigationActions,
  NavigationContainerComponent,
  NavigationParams,
  NavigationNavigateActionPayload
} from 'react-navigation'

let _navigator : NavigationContainerComponent | null

function setTopLevelNavigator(navigatorRef: NavigationContainerComponent | null) {
  _navigator = navigatorRef
}

function isNavigatePayload(obj: any): obj is NavigationNavigateActionPayload {
  return obj.routeName !== undefined 
}

function navigate(routeNameOrOptions: string | NavigationNavigateActionPayload, params?: NavigationParams) {
  
  let options: NavigationNavigateActionPayload = {
    routeName: ''
  }
  if (isNavigatePayload(routeNameOrOptions)) {
    options = routeNameOrOptions
  }
  else {
    options = {
      routeName: routeNameOrOptions
    }
  }
  if (params) {
    options.params = params
  }
  
  _navigator && _navigator.dispatch(
    NavigationActions.navigate(options)
  )
}

// add other navigation functions that you need and export them

export default {
  navigate,
  setTopLevelNavigator,
}