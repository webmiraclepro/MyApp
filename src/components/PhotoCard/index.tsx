import React, { useState, useEffect } from 'react'
import { TouchableOpacity, Image, View } from 'react-native'
import { SharedElement } from 'react-navigation-shared-element'
import { View as AnimatableView } from 'react-native-animatable'
import Icon from 'react-native-vector-icons/MaterialIcons'

import { SCREENS } from '../../constants'
import { Photo } from '../../models'
import { NavigationService, PhotoService } from '../../services'
import styles from './styles'

export function onItemPress(photo: Photo) {
  NavigationService.navigate(SCREENS.DETAILS, { data: photo })
}

const PhotoCard = (photo: Photo) => {
  const [downloaded, setDownload] = useState(true)

  const savePhoto = async (data: Photo) => {
    await PhotoService.addPhoto(data)
    setDownload(true)
  }

  useEffect(() => {
    PhotoService.getPhoto(photo.id).then((u) => {
      setDownload(!!u)
    })
  }, [photo.id])

  return (
    <View style={styles.item}>
      <TouchableOpacity
        testID="PhotoItem"
        key={photo.id}
        style={styles.card}
        onPress={() => onItemPress(photo)}>
        <AnimatableView
          useNativeDriver
          delay={2000}
          animation="pulse"
          easing="ease-out"
          style={styles.content}>
          <SharedElement id={photo.id} style={styles.fullSize}>
            <Image
              style={styles.fullSize}
              resizeMode="contain"
              source={{ uri: photo.photo }}
            />
          </SharedElement>
        </AnimatableView>
      </TouchableOpacity>
      <TouchableOpacity
        onPress={() => savePhoto(photo)}
        style={styles.downloadButton}
        disabled={downloaded}>
        <Icon
          name={downloaded ? 'cloud-done' : 'cloud-download'}
          size={30}
          color="black"
        />
      </TouchableOpacity>
    </View>
  )
}

export default PhotoCard
