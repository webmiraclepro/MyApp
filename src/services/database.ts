import { Database } from '@nozbe/watermelondb'
import SQLiteAdapter from '@nozbe/watermelondb/adapters/sqlite'

import schema from '../models/schema'
import migrations from '../models/migrations'
import Photo from '../models/photo'

// First, create the adapter to the underlying database:
const adapter = new SQLiteAdapter({
  schema,
  migrations,
})

// Then, make a Watermelon database from it!
const database = new Database({
  adapter,
  modelClasses: [Photo],
  actionsEnabled: true,
})

export default database
