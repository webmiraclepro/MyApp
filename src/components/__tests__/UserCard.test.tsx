import 'react-native'
import React from 'react'
import { mount, shallow } from 'enzyme'

import PhotoCard from '../PhotoCard'
import { NavigationService } from '../../services'

jest.mock('../../services', () => ({
  PhotoService: {
    getPhoto: jest.fn(() => Promise.resolve('mockedPhoto')),
    addPhoto: jest.fn(() => Promise.resolve()),
  },
  NavigationService: {
    navigate: jest.fn(),
  },
}))

describe('<PhotoCard />', () => {
  const defaultProps = {
    id: '123',
  } as any

  it('should render', () => {
    const rendered = mount(<PhotoCard {...defaultProps} />)
    expect(rendered).toBeTruthy()
  })

  it('renders correctly', () => {
    expect(shallow(<PhotoCard {...defaultProps} />)).toMatchSnapshot()
  })

  describe('When the card is pressed', () => {
    const user = { id: 123 } as any

    it('should redirect to the Details screen', () => {
      const wrapper = mount(<PhotoCard {...user} />)
      wrapper.find({ testID: 'PhotoItem' }).first().prop('onPress')()

      expect(NavigationService.navigate).toHaveBeenCalled()
    })
  })
})
