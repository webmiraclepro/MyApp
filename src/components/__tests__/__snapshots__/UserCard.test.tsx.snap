// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoCard /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "height": 200,
      "justifyContent": "center",
      "margin": 10,
      "position": "relative",
      "width": "100%",
    }
  }
>
  <ForwardRef
    key="123"
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "#ECECEC",
        "borderColor": "#CCC",
        "borderRadius": 5,
        "borderWidth": 0.5,
        "elevation": 3,
        "flex": 1,
        "flexDirection": "row",
        "padding": 10,
        "shadowColor": "#EEE",
        "shadowOffset": Object {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.4,
        "shadowRadius": 5,
      }
    }
    testID="PhotoItem"
  >
    <withAnimatable(View)
      animation="pulse"
      delay={2000}
      direction="normal"
      easing="ease-out"
      iterationCount={1}
      iterationDelay={0}
      onAnimationBegin={[MockFunction]}
      onAnimationEnd={[MockFunction]}
      onTransitionBegin={[MockFunction]}
      onTransitionEnd={[MockFunction]}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
      useNativeDriver={true}
    >
      <SharedElement
        id="123"
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "flex": 1,
            }
          }
        />
      </SharedElement>
    </withAnimatable(View)>
  </ForwardRef>
  <ForwardRef
    disabled={true}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 20,
        "bottom": 10,
        "elevation": 4,
        "height": 40,
        "justifyContent": "center",
        "position": "absolute",
        "width": 40,
        "zIndex": 2,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      color="black"
      name="cloud-done"
      size={30}
    />
  </ForwardRef>
</View>
`;
